nft-client/src/cts1.rs

```diff
@@ -0,0 +1,377 @@
+                owner,
+                operator,
+            } => {
+                let operation = match update {
+                    OperatorUpdate::Remove => "Remove",
+                    OperatorUpdate::Add => "Add",
+                };
+                write!(
+                    f,
+                    "{} {} as operator for {}",
+                    operation,
+                    address_display(operator),
+                    address_display(owner)
+                )?;
+            }
+            Event::TokenMetadata {
+                token_id,
+                metadata_url,
+            } => {
+                let hash = if let Some(hash) = metadata_url.hash {
+                    format!("with hash {}", hex::encode(hash))
+                } else {
+                    "without hash".to_string()
+                };
+                write!(
+                    f,
+                    "Added metadata url {} ({}) for token with ID {}",
+                    metadata_url.url, hash, token_id
+                )?;
+            }
+            Event::Unknown => {
+                write!(f, "Unknown event: Event is not part of CTS1 specification")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Deserialize the contract events as according to the CTS1 specification.
+impl Deserial for Event {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let discriminant = u8::deserial(source)?;
+        match discriminant {
+            0 => Ok(Event::Transfer {
+                token_id: TokenIdVec::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                from:     Address::deserial(source)?,
+                to:       Address::deserial(source)?,
+            }),
+            1 => Ok(Event::Mint {
+                token_id: TokenIdVec::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                owner:    Address::deserial(source)?,
+            }),
+            2 => Ok(Event::Burn {
+                token_id: TokenIdVec::deserial(source)?,
+                amount:   TokenAmount::deserial(source)?,
+                owner:    Address::deserial(source)?,
+            }),
+            3 => Ok(Event::UpdateOperator {
+                update:   OperatorUpdate::deserial(source)?,
+                owner:    Address::deserial(source)?,
+                operator: Address::deserial(source)?,
+            }),
+            4 => Ok(Event::TokenMetadata {
+                token_id:     TokenIdVec::deserial(source)?,
+                metadata_url: MetadataUrl::deserial(source)?,
+            }),
+            _ => Ok(Event::Unknown),
+        }
+    }
+}
+
+/// Display the Address using either the display for account address or contract
+/// address.
+fn address_display(a: &Address) -> String {
+    match a {
+        Address::Account(addr) => format!("{}", addr),
+        Address::Contract(addr) => format!("{}", addr),
+    }
+}
```

Comment:
Would it make sense to instead add a Display implementation to `Address`? This seems ad-hoc and hard to maintain.

Reply:
I think it would be nice if we provide this yes.
But it would have to go into `concordium_contracts_common` in this case.

We will have to decide how to display the contract address then.